File: src/Enterprise/Java/patterns/state.java
class PatternCell {
    enum GateState {
        LOCKED,
        UNLOCKED
    }

    enum Action {
        COIN,
        PUSH
    }

    static final class Turnstile {
        private GateState state = GateState.LOCKED;

        GateState state() {
            return state;
        }

        String apply(Action action) {
            return switch (state) {
                case LOCKED -> switch (action) {
                    case COIN -> {
                        state = GateState.UNLOCKED;
                        yield "unlocked";
                    }
                    case PUSH -> "blocked";
                };
                case UNLOCKED -> switch (action) {
                    case COIN -> "coin-returned";
                    case PUSH -> {
                        state = GateState.LOCKED;
                        yield "passed";
                    }
                };
            };
        }
    }

    private static void require(boolean condition) {
        if (!condition) {
            throw new AssertionError("Java State contract failed");
        }
    }

    public static void main(String[] args) {
        Turnstile gate = new Turnstile();
        require(gate.state() == GateState.LOCKED);

        require(gate.apply(Action.PUSH).equals("blocked"));
        require(gate.state() == GateState.LOCKED);

        require(gate.apply(Action.COIN).equals("unlocked"));
        require(gate.state() == GateState.UNLOCKED);

        require(gate.apply(Action.COIN).equals("coin-returned"));
        require(gate.state() == GateState.UNLOCKED);

        require(gate.apply(Action.PUSH).equals("passed"));
        require(gate.state() == GateState.LOCKED);

        System.out.println("java-state: passed");
    }
}

File: src/Functional/Haskell/State.hs
module Main where

data GateState = Locked | Unlocked
  deriving (Eq, Show)

data GateAction = InsertCoin | Push
  deriving (Eq, Show)

data GateResult = CoinAccepted | CoinReturned | Passed | Blocked
  deriving (Eq, Show)

transition :: GateState -> GateAction -> (GateState, GateResult)
transition Locked InsertCoin = (Unlocked, CoinAccepted)
transition Locked Push = (Locked, Blocked)
transition Unlocked InsertCoin = (Unlocked, CoinReturned)
transition Unlocked Push = (Locked, Passed)

must :: Bool -> IO ()
must True = pure ()
must False = error "State contract failed"

main :: IO ()
main = do
  let initial = Locked
      (afterBlockedPush, blockedResult) = transition initial Push
      (afterCoin, coinResult) = transition afterBlockedPush InsertCoin
      (afterDuplicateCoin, duplicateCoinResult) = transition afterCoin InsertCoin
      (afterPass, passResult) = transition afterDuplicateCoin Push

  must (initial == Locked)
  must (afterBlockedPush == Locked && blockedResult == Blocked)
  must (afterCoin == Unlocked && coinResult == CoinAccepted)
  must (afterDuplicateCoin == Unlocked && duplicateCoinResult == CoinReturned)
  must (afterPass == Locked && passResult == Passed)

  putStrLn "haskell-state: passed"
